include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -265,238 +308,125 @@ vec4 readReg4(in uint n)
-**/
+	uvec3 r = instr_decodeRegisters(i);
+	return instr_fetchSrcRegs(i, r);
+}
 
-	vec3 H; // @Crisspl its your reponsibility not to evaluate/execute (then) if irr_glsl_isTransmissionPath==true and either the BSDF is not tranmissive, or the microfacet normal is invalid
-	irr_glsl_BSDFIsotropicParams isoparams = irr_glsl_calcBSDFIsotropicParams(irr_glsl_isTransmissionPath(currInteraction.isotropic.NdotV,NdotL), L, , currInteraction.isotropic, H);
-	currBSDFParams = irr_glsl_calcBSDFAnisotropicParams(L, H, isoparams, currInteraction);
+void setCurrInteraction(in vec3 N, in vec3 V, in vec3 pos)
+{
+	irr_glsl_IsotropicViewSurfaceInteraction interaction = irr_glsl_calcFragmentShaderSurfaceInteractionFromViewVector(V, pos, N);
+	currInteraction = irr_glsl_calcAnisotropicInteraction(interaction);
+}
+void setCurrInteraction(in MC_precomputed_t precomp)
+{
+	setCurrInteraction(precomp.NdotV > 0.0 ? precomp.N : -precomp.N, precomp.V, precomp.pos);
+}
+void updateCurrInteraction(in MC_precomputed_t precomp, in vec3 N)
+{
+	setCurrInteraction(precomp.NdotV > 0.0 ? N : -N, precomp.V, precomp.pos);
 }
 
+//clamping alpha to min MIN_ALPHA because we're using microfacet BxDFs for deltas as well (and NDFs often end up NaN when given alpha=0) because of less deivergence
+//probably temporary solution
+#define MIN_ALPHA 0.0001
 float params_getAlpha(in params_t p)
 {
-	return p[PARAMS_ALPHA_U_IX].x;
+	return max(p[PARAMS_ALPHA_U_IX].x,MIN_ALPHA);
 }
 vec3 params_getReflectance(in params_t p)
 {
 	return p[PARAMS_REFLECTANCE_IX];
 }
-vec3 params_getOpacity(in params_t p)
-{
-	return p[PARAMS_OPACITY_IX];
-}
 float params_getAlphaV(in params_t p)
 {
-	return p[PARAMS_ALPHA_V_IX].x;
+	return max(p[PARAMS_ALPHA_V_IX].x,MIN_ALPHA);
 }
 vec3 params_getSigmaA(in params_t p)
 {
 	return p[PARAMS_SIGMA_A_IX];
 }
-float params_getBlendWeight(in params_t p)
+vec3 params_getBlendWeight(in params_t p)
 {
-	return p[PARAMS_WEIGHT_IX].x;
+	return p[PARAMS_WEIGHT_IX];
 }
 vec3 params_getTransmittance(in params_t p)
 {
 	return p[PARAMS_TRANSMITTANCE_IX];
 }
 
-
-#ifdef OP_DIFFUSE
-void instr_execute_cos_eval_DIFFUSE(in instr_t instr, in uvec3 regs, in params_t params, in bsdf_data_t data)
+bxdf_eval_t instr_execute_cos_eval_COATING(in instr_t instr, in mat2x4 srcs, in params_t params, in vec3 eta, in vec3 eta2, in irr_glsl_LightSample s, in irr_glsl_AnisotropicMicrofacetCache microfacet, in bsdf_data_t data, out float out_weight)
 {
-	if (currBSDFParams.isotropic.NdotL>FLT_MIN)
-	{
-		vec3 refl = params_getReflectance(params);
-		float a = params_getAlpha(params);
-		vec3 diffuse = irr_glsl_oren_nayar_cos_eval(currBSDFParams.isotropic,currInteraction.isotropic,a*a) * refl;
-		writeReg(REG_DST(regs), diffuse);
-	}
-	else writeReg(REG_DST(regs), bxdf_eval_t(0.0));
-}
-#endif
+	//vec3 thickness_sigma = params_getSigmaA(params);
+	
+	vec3 ws = irr_glsl_fresnel_dielectric_frontface_only(eta, microfacet.isotropic.VdotH);
```

Comment:
we expect that the COAT has already computed the fresnel for itself (dont want double application of fresnel)

so `ws` shouldnt exist and the computation at the end should be `coat+coated*wd`